core/include/edm/container.hpp

```diff
@@ -64,7 +66,8 @@ class container_element {
 /// header and item type.
 template <typename header_t, typename item_t,
           template <typename> class vector_t,
-          template <typename> class jagged_vector_t>
+          template <typename> class jagged_vector_t,
+          template <typename...> class pair_type = std::pair>
```

Comment:
Do you foresee using something other than [std::pair](https://en.cppreference.com/w/cpp/utility/pair) here? Like [std::tuple](https://en.cppreference.com/w/cpp/utility/tuple) let's say?

I'm not completely sold even that we need to template this type at all. But as long as we only want to choose between [std::pair](https://en.cppreference.com/w/cpp/utility/pair) and [thrust::pair](https://thrust.github.io/doc/structthrust_1_1pair.html), you should declare this type rather like:

```c++
template <typename, typename> class pair_type = std::pair
```

I'm also not too happy with the naming here. As we discussed before, I myself would prefer an implementation like:

```c++
template <typename header_t, typename item_t,
          template <typename> class vector_t,
          template <typename> class jagged_vector_t,
          template <typename, typename> class pair_t = std::pair>
class container {
    ...
    template <typename T1, typename T2>
    using pair_type = pair_t<T1, T2>;
    ...
```

Just so that the `container::pair_type` symbol could be used outside of the class's body as well.

P.S. If for some reason you do want to consider [std::tuple](https://en.cppreference.com/w/cpp/utility/tuple) then you must replace all uses of `first` and `second`, and rather access the pair's elements with [std::get](https://en.cppreference.com/w/cpp/utility/pair/get) everywhere. But I don't think that should be necessary...

Reply:
It's only for pair - I changed to `template <typename, typename> class pair_t = std::pair`